janus_collector/src/lib.rs

```diff
@@ -0,0 +1,866 @@
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[non_exhaustive]
+pub enum Authentication {
+    /// Beearer token authentication, via the `DAP-Auth-Token` header.
+    BearerToken(#[derivative(Debug = "ignore")] String),
+}
+
+/// The PPM collector's view of task parameters.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct CollectorParameters {
+    /// Unique identifier for the task.
+    task_id: TaskId,
+    /// The base URL of the leader's aggregator API endpoints.
+    #[derivative(Debug(format_with = "std::fmt::Display::fmt"))]
+    leader_endpoint: Url,
+    /// The authentication information needed to communicate with the leader aggregator.
+    authentication: Authentication,
+    /// HPKE configuration and public key used for encryption of aggregate shares.
+    hpke_config: HpkeConfig,
+    /// HPKE private key used to decrypt aggregate shares.
+    #[derivative(Debug = "ignore")]
+    hpke_private_key: HpkePrivateKey,
+    /// Parameters to use when retrying HTTP requests.
+    http_request_retry_parameters: ExponentialBackoff,
+    /// Parameters to use when waiting for a collect job to be processed.
+    collect_poll_wait_parameters: ExponentialBackoff,
+}
+
+impl CollectorParameters {
+    /// Creates a new set of collector task parameters.
+    pub fn new(
+        task_id: TaskId,
+        leader_endpoint: Url,
+        authentication_token: String,
+        hpke_config: HpkeConfig,
+        hpke_private_key: HpkePrivateKey,
+    ) -> CollectorParameters {
+        CollectorParameters::new_with_backoff(
+            task_id,
+            leader_endpoint,
+            authentication_token,
+            hpke_config,
+            hpke_private_key,
+            http_request_exponential_backoff(),
+            ExponentialBackoff {
+                initial_interval: StdDuration::from_secs(15),
+                max_interval: StdDuration::from_secs(300),
+                multiplier: 1.5,
+                max_elapsed_time: None,
+                ..Default::default()
+            },
+        )
+    }
+
+    /// Creates a new set of collector task parameters with non-default HTTP request retry
+    /// parameters.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new_with_backoff(
+        task_id: TaskId,
+        mut leader_endpoint: Url,
+        authentication_token: String,
+        hpke_config: HpkeConfig,
+        hpke_private_key: HpkePrivateKey,
+        http_request_retry_parameters: ExponentialBackoff,
+        collect_poll_wait_parameters: ExponentialBackoff,
+    ) -> CollectorParameters {
+        // Ensure the provided leader endpoint ends with a slash, as we will be joining additional
+        // path segments into this endpoint. The Url::join implementation will drop the last
+        // path component of the base URL if it does not end with a slash.
+        if !leader_endpoint.as_str().ends_with('/') {
+            leader_endpoint.set_path(&format!("{}/", leader_endpoint.path()));
+        }
```

Comment:
This is our third copy of this URL sanitization code, after `janus_client/src/lib.rs` and `janus_server/src/task.rs`. I think it's time to put something like `fn sanitize_url()` in `janus_core`.

Reply:
Alternatively, maybe we should just have our own `actually_join_url()` with correct semantics, and enforce that it's called instead of `Url::join` everywhere (via auditing/code review; static analysis would work too but is probably too much effort).

The upside (IMO) is that it's a lot easier to audit for a particular method (not) being used than that every path to storing a URL happens to call a `sanitize_url` method.